src/main/java/com/example/demo/StudentService.java

```diff
@@ -16,27 +16,17 @@ public List<Student> getAllStudents() {
     }
 
     public Student getStudentById(String id) {
-        if (studentRepository.existsById(id)) {
-            Optional<Student> student = studentRepository.findById(id);
+            Optional<Student> student = studentRepository.getById(id);
             return student.orElse(null);
-        }
-        return null;
     }
 
     public Student updateStudent(String id, Student studentDetails) {
-        if (studentRepository.existsById(id)) {
             return studentRepository.save(studentDetails);
```

Comment:
I'm guessing that we want to separate this sort of behavior so that it all occurs within the repository layer. Since save doesn't currently have an implementation in the repository, I would assume that eventually it will be the one to handle whether or not the id is valid. Would it just return a null for the case where it is not a valid id and then the controller would have to interpret that null? I guess I have multiple ideas and questions for this.

Reply:
Checking if ids exist will belong in the service layer. 

The repository should be only a medium to communicate with the Database. All validity checks should happen in the service layer. 

And I would prefer that it returns an java.util.Optional. This way the service can determine out to act without using nulls. Works with nulls should be a last result if we can help it.

Reply:
I'm not sure what you mean by "All validity checks should happen in the service layer.". Isn't communication with the DB required to determine whether or not an id exists within the DB? Also, in your change, you deleted the line that checked if an id exists. Is this because it used the studentRepository method to accomplish this? 

Do you mean to return an Optional as we see in the getByID method of StudentRepository? Or did you have something else in mind?

Reply:
Yes. 
No, I deleted it because I just needed to get the project in a build-able state. 

Yes.

Reply:
Now let me answer: "What do you mean by All Validity checks should happen in the service layer?"

What I mean is, the repository methods should only do one thing for each method. Such as getting a single user, or a list of users. Nothing more than that. Then, in the service layer, you use the results of the repository method to base decisions make decisions on how the application should behave. For example: If the client submits a create user request, the service should attempt to retrieve a user with those properties from the DB using a "getUserBy.." method. If a User does already exist, the service should then decide to use the "update" repository method, instead of the "create".

Reply:
Okay, so if I wanted to get a single user by id, I would send the request over to the repository and it would return an optional user response. In the case where that user didn't exist, it wouldn't return a viable user object so the service would then interpret that as that user not existing. The repository is supposed to perform any of the 'thinking', it only does simple, exactly as it is told, sort of operations. The service layer is tasked with interpreting those responses.

In your example of creating a user, why would you choose the update method when the user already exists? Why not return a response to the client that the user already exists? Also, what metrics determine whether you're actually trying to create a new user vs. recreating a user that has already been made, but with slightly different parameters? I know that you could compare across all variables to determine if the objects are truly the same, but what is the best way to differentiate at a less-encompassing level? I know that there are ways of making columns unique in SQL, is that how you would go about doing it? We can only have one entry with that email, so your request is denied.

Reply:
Paragraph 1: Correct.

Paragraph 2:
1.  Returning a response to the client that the user already exists is an acceptable choice. 
2. The client will tell the server of it's intended expectations by endpoint. If it wants to create a user, it will hit the create user endpoint, regardless of payload to the endpoint.
3. My preference is twofold: The DB enforces the constraint, and the service layer should protect the application from hitting that constraint.